Sources/StreamChatUI/Utils/Throttler.swift

```diff
@@ -0,0 +1,54 @@
+//
+// Copyright © 2023 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+/// A throttler implementation. The action provided will only be executed if the last action executed has passed an amount of time.
+/// Based on the implementation from Apple: https://developer.apple.com/documentation/combine/anypublisher/throttle(for:scheduler:latest:)
+class Throttler {
+    private var workItem: DispatchWorkItem?
+    private let queue = DispatchQueue(label: "com.stream.throttler", qos: .background)
+    private var previousRun: Date = Date.distantPast
+    let interval: TimeInterval
+    let broadcastLatestEvent: Bool
+
+    /// - Parameters:
+    ///   - interval: The interval that an action can be executed.
+    ///   - broadcastLatestEvent: A Boolean value that indicates whether we should be using the first or last event of the ones that are being throttled.
+    ///   This last action will have a delay of the provided interval until it is executed.
+    init(interval: TimeInterval, broadcastLatestEvent: Bool = true) {
+        self.interval = interval
+        self.broadcastLatestEvent = broadcastLatestEvent
+    }
+
+    /// Throttle an action. It will cancel the previous action if exists, and it will execute the action immediately
+    /// if the last action executed was past the interval provided. If not, it will only be executed after a delay.
+    /// - Parameter action: The closure to be performed.
+    func throttle(_ action: @escaping () -> Void) {
+        workItem?.cancel()
+
+        let workItem = DispatchWorkItem { [weak self] in
+            guard let workItem = self?.workItem, !workItem.isCancelled else { return }
+            action()
+            self?.previousRun = Date()
+            self?.workItem = nil
+        }
+
+        self.workItem = workItem
```

Comment:
That seems more like a debouncer rather a throttle as you always have one item that is being executed after some interval (and the ones in the middle are being discarded).
I'm using [this](https://web.archive.org/web/20180324022838/http://demo.nimius.net/debounce_throttle/) visualisation to separate those 2.

If you end up needing a debouncer instead of a throttler we already have an implementation in LLC. If you continue with the throttler we will need some tests for each functionality.

Reply:
On the tests side of things, we have decided not to go with them for now, until all the assumptions are validated.

On the throttling, it looks quite similar to the debouncing, but this changes it all: https://github.com/GetStream/stream-chat-swift/pull/2808/files#diff-b726cf389a7cfaa48b83a83145188d57562aff5f1eaa6bc8dcb98e1fdc513eedR41

Reply:
This one is an awesome article to understand better the difference between Throttling and Debouncing: https://redd.one/blog/debounce-vs-throttle

This class was done based on that as well